diff --git a/src/elan-dist/src/component/package.rs b/src/elan-dist/src/component/package.rs
index 89ff5ce..d412dc1 100644
--- a/src/elan-dist/src/component/package.rs
+++ b/src/elan-dist/src/component/package.rs
@@ -51,6 +51,26 @@ fn unpack_without_first_dir<R: Read>(archive: &mut tar::Archive<R>, path: &Path)
         entry
             .unpack(&full_path)
             .chain_err(|| ErrorKind::ExtractingPackage)?;
+        nix_patch_if_needed(&full_path)?;
+    }
+
+    Ok(())
+}
+
+fn nix_patch_if_needed(dest_path: &Path) -> Result<()> {
+    let is_bin = matches!(dest_path.parent(), Some(p) if p.ends_with("bin"));
+    if !is_bin {
+        return Ok(());
+    }
+    let _ = ::std::process::Command::new("@patchelf@/bin/patchelf")
+        .arg("--set-interpreter")
+        .arg("@dynamicLinker@")
+        .arg(dest_path)
+        .output();
+
+    if dest_path.file_name() == Some(::std::ffi::OsStr::new("llvm-ar")) {
+        ::std::fs::remove_file(dest_path)?;
+        ::std::os::unix::fs::symlink("@ar@", dest_path)?;
     }
 
     Ok(())
